stacks/api/machinery/src/validation.ts

```diff
@@ -0,0 +1,82 @@
+import {
+  ErrorObject,
+  ValidationError as _ValidationError,
+  Validator,
+  _
+} from "@spica-server/core/schema";
+import {Resource} from "./definition";
+import {getVersionFromScheme, Scheme} from "./scheme";
+import {badRequest} from "./status";
+
+export interface ValidationOptions {
+  scheme: Scheme;
+  versionName: string;
+  object: Resource<object, unknown>;
+}
+
+export interface DataError {
+  path: string;
+  message: string;
+}
+
+export async function validate({object, scheme, versionName}: ValidationOptions): Promise<void> {
+  const version = getVersionFromScheme(scheme, versionName);
+  const validator = new Validator({schemas: version.additionalSchemas}, {});
+  try {
+    await validator.validate(version.schema, object.spec);
+  } catch (e) {
+    if (e instanceof _ValidationError) {
+      throw badRequest({
+        message: "Validation error",
+        reason: "Invalid",
+        details: {
+          errors: e.errors.map(mapErrorObjectToDataError)
+        }
+      });
+    }
+
+    throw badRequest({
+      message: e.message,
+      reason: "Invalid",
+      details: {
+        error: e,
+        params: e.constructor.name
+      }
+    });
+  }
+}
+
+// function groupErrors(errors: Partial<ErrorObject>[]): Partial<ErrorObject>[] {
+//     const oneOfIndex = errors.findIndex(error => error.keyword == "oneOf" );
+//     if ( oneOfIndex != -1 ) {
+//         const oneOf = errors[oneOfIndex];
+//         const oneOfRelatedErrors = errors.filter((error, index) => error.dataPath == oneOf.dataPath && index != oneOfIndex);
+//         if (! oneOf.params.passingSchemas ) {
+//             oneOf.message = `either ${oneOfRelatedErrors.map(error => error.message).join(" or ")}`;
+//             errors = errors.filter((error, index) => error.dataPath != oneOf.dataPath || index == oneOfIndex);
+//         } else {
+//             console.log(oneOf);
+//         }
+
+//     }
+
+//     return errors;
+// }
```

Comment:
Can we add one more comment line that explains why we disabled these lines or uncomment all of them?

Reply:
This was a part of an experimental feature that I have failed to implement. Better remove it

Reply:
Done